src/Core/ConstraintValidator/ZipCodeRangeValidator.php

```diff
@@ -0,0 +1,123 @@
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\ConstraintValidator;
+
+use PrestaShop\PrestaShop\Core\CommandBus\CommandBusInterface;
+use PrestaShop\PrestaShop\Core\Domain\Country\Exception\CountryConstraintException;
+use PrestaShop\PrestaShop\Core\Domain\Country\Query\GetCountryZipCodeRequirements;
+use PrestaShop\PrestaShop\Core\Domain\Country\QueryResult\CountryZipCodeRequirements;
+use PrestaShop\PrestaShop\Core\Form\FormChoiceProviderInterface;
+use Symfony\Component\Translation\TranslatorInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+
+/**
+ * Validate zip codes in range
+ */
+class ZipCodeRangeValidator extends ConstraintValidator
+{
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var CommandBusInterface
+     */
+    private $queryBus;
+
+    /**
+     * @var FormChoiceProviderInterface
+     */
+    private $countryChoiceProvider;
+
+    /**
+     * @param TranslatorInterface $translator
+     * @param CommandBusInterface $queryBus
+     * @param FormChoiceProviderInterface $countryChoiceProvider
+     */
+    public function __construct(
+        TranslatorInterface $translator,
+        CommandBusInterface $queryBus,
+        FormChoiceProviderInterface $countryChoiceProvider
+    ) {
+        $this->translator = $translator;
+        $this->queryBus = $queryBus;
+        $this->countryChoiceProvider = $countryChoiceProvider;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws CountryConstraintException
+     */
+    public function validate($value, Constraint $constraint)
+    {
+        $countryId = (int) $value['country'];
+        $selectedCountries = [$countryId];
+        $zipCode = $value['zipCode'];
+
+        if (null === $zipCode) {
+            return;
+        }
+
+        if ($countryId === 0) {
+            $countries = $this->countryChoiceProvider->getChoices();
+            $selectedCountries = array_map('intval', $countries);
+        }
+
+        foreach ($selectedCountries as $country) {
+            /** @var CountryZipCodeRequirements $requirements */
+            $requirements = $this->queryBus->handle(new GetCountryZipCodeRequirements($country));
+
+            foreach (explode('-', $zipCode) as $code) {
```

Comment:
We could have two separate textboxes for `from` and `to` values separated by `-`. In this case these could be VO's but it would change form design a bit. Would that be ok?

Reply:
No, let's keep the design like it is.
If I understand correctly: in the view, there is a field where someone can put multiple zipcodes, separated by `-`. So the view is like this, we do not modify it.
However when the controller retrieves the view data, it should convert this string into an array of zipcodes. The data that travels between the controller and the domain should be as accurate as possible. Si if we are manipulating an array of zipcodes, let's manipulate a real `array` instead of a string-that-contains-concatenated-zipcodes.

Is that good ? 😉

Reply:
Legacy can take multiple many zipcodes separated by `-` but only parses the first two. So there is always going to be a zip code array of one or two elements. Maybe `ZipCodeDataTransformer` would be a good idea to use here?